src/openms/source/FILTERING/DATAREDUCTION/Deisotoper.cpp

```diff
@@ -38,9 +38,292 @@
+
+  if (spec.empty()) { return; }
+
+  // discard low-intensity peaks
+  if (rem_low_intensity)
+  { 
+    Size max_num_peaks = used_for_open_search ? 1000 : 5000;
+
+    spec.sortByIntensity();
+    if (spec.size() > max_num_peaks || spec[0].getIntensity() == 0.0)
+    {
+      // establish intensity that is not accepted anymore
+      Peak1D::IntensityType unacceptable = (spec.size() > max_num_peaks) ? spec[spec.size() - max_num_peaks - 1].getIntensity() : 0;
+
+      std::vector<Size> idx;
+      idx.reserve(max_num_peaks); // prevent foreseeable reallocations
+      
+      for (Size spec_idx = spec.size() - 1; spec[spec_idx].getIntensity() > unacceptable; --spec_idx)
+      {
+        idx.push_back(spec_idx);
+      }
+
+      spec.select(idx);
+    }
+    spec.sortByPosition();
+  }
+
+  Size charge_index(0);
+  Size iso_peak_count_index(0);
+
+  // reserve integer data array to store charge of peaks
+  if (annotate_charge)
+  {
+    // expand to hold one additional integer data array to hold the charge
+    spec.getIntegerDataArrays().resize(spec.getIntegerDataArrays().size() + 1);
+    spec.getIntegerDataArrays().back().setName("charge");
+    charge_index = spec.getIntegerDataArrays().size() - 1;
+  }
+  // reserve integer data array to store number of isotopic peaks for each isotopic pattern
+  if (annotate_iso_peak_count)
+  {
+    spec.getIntegerDataArrays().resize(spec.getIntegerDataArrays().size() + 1);
+    spec.getIntegerDataArrays().back().setName("iso_peak_count");
+    iso_peak_count_index = spec.getIntegerDataArrays().size() - 1;
+  }
+
+  // during discovery phase, work on a constant reference (just to make sure we do not modify spec)
+  const MSSpectrum& old_spectrum = spec;
+
+  // determine charge seeds and extend them
+  std::vector<size_t> mono_isotopic_peak(old_spectrum.size(), 0);
+  std::vector<int> features(old_spectrum.size(), -1);
+  std::vector<double> mono_iso_peak_intensity(old_spectrum.size(), 0);
+  std::vector<Size> iso_peak_count(old_spectrum.size(), 1);
+  int feature_number = 0;
+
+  std::vector<size_t> extensions;
+  std::vector< std::vector<size_t> > clusters;
+  std::vector<int> charges_of_extensions;
+  const float averagine_check_threshold[5] = { 0.05f,0.1f,0.2f,0.4f,0.6f };
+
+  bool has_precursor_data(false);
+  double precursor_mass(0);
+  if (old_spectrum.getPrecursors().size() == 1)
+  {
+    has_precursor_data = true;
+    int precursor_charge = old_spectrum.getPrecursors()[0].getCharge();
+    precursor_mass = (old_spectrum.getPrecursors()[0].getMZ() * precursor_charge) - (Constants::PROTON_MASS * precursor_charge);
+  }
+
+  for (size_t current_peak = 0; current_peak != old_spectrum.size(); ++current_peak)
+  {
+    // only process peaks which are not already in a cluster. Would form clusters identical to the cluster they are assigned to,
+    // excluding its first peak(s), since peaks with lower mz are not considered for cluster-formation.
+    if (features[current_peak] != -1) continue;
```

Comment:
note to myself: check in publication.
e.g., sometimes with deamidation it might make sense to take a smaller cluster with better KL

Reply:
can you check that this is correct?

Reply:
The paper does not explicitly say if peaks already in a cluster should be examined again. In my experience e.g. deamination peaks do not pass the averagine check anyway, as is tested as well:
https://github.com/OpenMS/OpenMS/blob/1b2aaa4d5e01376df3b9d002a4178c9764a9632c/src/tests/class_tests/openms/source/Deisotoper_test.cpp#L187

I do not believe that changing this would alter the output, as the averagine check should only be passed by the correct cluster, but it might be worth a try. Will give it a look.